packages/flutter/lib/src/widgets/router.dart

```diff
@@ -0,0 +1,1327 @@
+        FlutterError.reportError(
+          const FlutterErrorDetails(
+            exception:
+              'Router.routeInformationParser returns a null RouteInformation. '
+              'If you opt for route information reporting, the '
+              'routeInformationParser must not report null for a given '
+              'configuration.'
+          ),
+        );
+      }
+      return true;
+    }());
+    return routeInformation;
+  }
+
+  void _setStateWithExplicitReportStatus(
+    _IntentionToReportRouteInformation status,
+    VoidCallback fn,
+  ) {
+    assert(status != null);
+    assert(status.index >= _IntentionToReportRouteInformation.must.index);
+    assert(() {
+      if (_currentIntentionToReporting.index >= _IntentionToReportRouteInformation.must.index &&
+          _currentIntentionToReporting != status) {
+        FlutterError.reportError(
+          const FlutterErrorDetails(
+            exception:
+              'Both Router.navigate and Router.neglect have been called in this '
+              'build cycle, and the Router cannot decide whether to report the '
+              'route information. Please make sure only one of them is called '
+              'within the same build cycle.'
+          ),
+        );
+      }
+      return true;
+    }());
+    _currentIntentionToReporting = status;
+    setState(fn);
+  }
+
+  void _maybeNeedToReportRouteInformation() {
+    _currentIntentionToReporting = _currentIntentionToReporting != _IntentionToReportRouteInformation.none
+      ? _currentIntentionToReporting
+      : _IntentionToReportRouteInformation.maybe;
+  }
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    _maybeNeedToReportRouteInformation();
+  }
+
+  @override
+  void setState(VoidCallback fn) {
+    super.setState(fn);
+    _maybeNeedToReportRouteInformation();
+  }
+
+  @override
+  void didUpdateWidget(Router<T> oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (widget.routeInformationProvider != oldWidget.routeInformationProvider ||
+        widget.backButtonDispatcher != oldWidget.backButtonDispatcher ||
+        widget.routeInformationParser != oldWidget.routeInformationParser ||
+        widget.routerDelegate != oldWidget.routerDelegate) {
+      _currentRouteInformationParserTransaction = Object();
+      _currentRouterDelegateTransaction = Object();
+    }
+    if (widget.routeInformationProvider != oldWidget.routeInformationProvider) {
+      oldWidget.routeInformationProvider
+        ?.removeListener(_handleRouteInformationProviderNotification);
+      widget.routeInformationProvider
+        ?.addListener(_handleRouteInformationProviderNotification);
+      if (oldWidget.routeInformationProvider?.value != widget.routeInformationProvider?.value) {
+        _handleRouteInformationProviderNotification();
+      }
+    }
+    if (widget.backButtonDispatcher != oldWidget.backButtonDispatcher) {
+      oldWidget?.backButtonDispatcher
```

Comment:
Why the `?` here? If `oldWidget` is null, it would already crash in the previous line when accessing its `backButtonDispatcher` there.